src/mscorlib/src/System/StubHelpers.cs

```diff
@@ -125,6 +125,112 @@ static internal void ClearNative(IntPtr pNative)
         }
     }  // class CSTRMarshaler
 
+    [ReliabilityContract(Consistency.WillNotCorruptState, Cer.MayFail)]
+    internal static class UTF8Marshaler
+    {
+        const int UTF8_CHAR_SIZE = 3;
+        static internal unsafe IntPtr ConvertToNative(int flags, string strManaged, IntPtr pNativeBuffer)
+        {
+            if (null == strManaged)
+            {
+                return IntPtr.Zero;
+            }
+            StubHelpers.CheckStringLength(strManaged.Length);
+
+            int nb;
+            byte* pbNativeBuffer = (byte*)pNativeBuffer;
+
+            // If we are marshaling into a stack buffer allocated by the ILStub
+            // we will use a "1-pass" mode where we convert the string directly into the unmanaged buffer.   
+            // else we will allocate the precise native heap memory.          
+            if (pbNativeBuffer != null)
+            {
+                // this is the number of bytes allocated by the ILStub.
+                nb = (strManaged.Length + 1) * UTF8_CHAR_SIZE;
+
+                // +1 for the '\0' that we add
+                nb += 1;
+
+                // nb is the actual number of bytes written by Encoding.GetBytes.
+                // use nb to de-limit the string since we are allocating more than 
+                // required on stack
+                nb = strManaged.GetBytesFromEncoding(pbNativeBuffer, nb, Encoding.UTF8);
+            }
+            // required bytes > 260 , allocate required bytes on heap             
+            else
+            {
+                nb = Encoding.UTF8.GetByteCount(strManaged);
```

Comment:
Given that most strings that we are dealing with will be in the ANSI range  , one option is to optimize for that case. ie allocate string.Length bytes. This is good as long as the string chars are in ANSI range , but the UTF8.GetBytes(..) will throw if the buffer is not big enough. Ideally if UTF8.GetBytes return the actual required bytes , then we can try again with the correct buffer size. But again it's a choice that we have to make , since it might not the best for non-ANSI chars.